api/src/main/java/org/openmrs/module/ohricore/api/impl/HIVStatusComputedConcept.java

```diff
@@ -0,0 +1,72 @@
+package org.openmrs.module.ohricore.api.impl;
+
+import org.openmrs.Concept;
+import org.openmrs.Encounter;
+import org.openmrs.EncounterType;
+import org.openmrs.Obs;
+import org.openmrs.Patient;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.ohricore.api.OHRIComputedConcept;
+import org.openmrs.module.ohricore.engine.CommonsUUID;
+import org.openmrs.module.ohricore.engine.HIVStatusConceptUUID;
+import org.springframework.stereotype.Component;
+
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Date;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+/**
+ * @author MayanjaXL, Amos, Stephen, smallGod date: 28/06/2021
+ */
+@Component("hivStatusComputedConcept")
+public class HIVStatusComputedConcept implements OHRIComputedConcept {
+	
+	@Override
+	public Obs compute(Encounter triggeringEncounter) {
+		
+		Concept hivFinalTestConcept = getConcept(HIVStatusConceptUUID.FINAL_HIV_TEST_RESULT);
+		
+		Patient patient = triggeringEncounter.getPatient();
+		List<Obs> hivTestObs = Context.getObsService().getObservationsByPersonAndConcept(patient.getPerson(),
+		    hivFinalTestConcept);
+		
+		Concept hivStatus = computeHivStatusConcept(hivTestObs);
+		return createOrUpdateObs(patient, hivStatus);
+	}
+	
+	private Concept computeHivStatusConcept(List<Obs> hivTestObs) {
+
+        Supplier<Stream<Obs>> hivTestObsStream  = hivTestObs::stream;
+        return hivTestObsStream.get()
+                .filter(obs -> obs.getValueCoded() == getConcept(CommonsUUID.POSITIVE))
+                .findAny()
+                .map(Obs::getValueCoded)
+                .orElse(hivTestObsStream.get()
+                        .filter(obs -> obs.getValueCoded() == getConcept(CommonsUUID.NEGATIVE))
+                        .filter(obs -> valueDateIsWithin90Days(obs.getValueDate()))
+                        .findAny()
+                        .map(Obs::getValueCoded)
+                        .orElse(getConcept(CommonsUUID.UNKNOWN))
+                );
+    }
+	
+	@Override
+	public EncounterType getTargetEncounterType() {
+		return null;
+	}
+	
+	@Override
+	public Concept getConcept() {
+		return Context.getConceptService().getConceptByUuid(HIVStatusConceptUUID.HIV_STATUS);
+	}
+	
+	private boolean valueDateIsWithin90Days(Date valueDate) {
```

Comment:
Change this method to valueDateWithinPeriod(String conceptUuid, Period, PeriodAmount) - check the Reporting module has a good approach for this or even the Calculation module

Reply:
@ssmusoke I looked at various implementations. I went on to have this implementation below. You can kindly check if it serves the purpose. I made it more generic to cater for different scenarios

      /**
     * Check if a given date falls within a given period/duration from Now
     *
     * @param dateToCheck  the date to check if it falls within a certain period/duration
     * @param periodUnit   Can be hours, days, years etc
     * @param periodAmount Period amount is negative if in the past or Positive if in the
     *                     future
     * @return true if given date is within duration, otherwise false
     */
     public static boolean dateWithinPeriodFromNow(Date dateToCheck, ChronoUnit periodUnit, int periodAmount) {

            LocalDateTime targetDateTime = LocalDateTime.now().plus(periodAmount, periodUnit);
            LocalDateTime dateTimeToCheck = LocalDateTime.ofInstant(dateToCheck.toInstant(), ZoneId.systemDefault());
            return dateTimeToCheck.isAfter(targetDateTime);
     }

Reply:
perhaps the other question is, in your parameters I saw this "String conceptUuid" Was that a mistake?

Reply:
@smallgod Nope it is not the type is a String and the variable Name is conceptUuid

Reply:
okay, @ssmusoke so I kindly needed to pick your mind on that. How can I use it in this context? Say after passing it to the method. What I am after is the date of observation value. How would I get it starting from a conceptUuid? Do I have to call the concept service to pick a concept? if yes, what happens after I have the concept?
Please excuse me if I am missing something